### ISSUE TITLE
Mimicking egrad using jacfwd and vmap 

### ISSUE THREAD

**Issue by leakec** [google/jax#3556]
Hello,

I am trying to mimic the use of egrad from autograd by composing vmap and jacfwd. This works well for simple functions, but I've had some strange shape behavior as the functions become slightly more complex. For example,

```python
#Import JAX
from jax.config import config
config.update('jax_enable_x64', True)

import jax.numpy as np
from jax import core, grad, vmap, abstract_arrays, jacfwd, jacrev

# Create simple data
x = np.arange(10,dtype=np.float64)
xi = np.ones(3)

A = np.ones((10,3))

# Create three functions, increasing in complexity
f = lambda x: np.array([x,x**2,x**3]).T
f2 = lambda x,xi: x*np.dot(f(x),xi)
f3 = lambda x,xi: x*np.dot(f(x),xi)+np.dot(A,xi)

# Attempt to mimic egrad behavior
def egrad(g):
    return vmap(jacfwd(g))
def egrad2(g):
    return vmap(jacfwd(g,0),(0,None),0)

# Print results
print("Function f")
print(egrad(f)(x))
print("")
print("Function f2")
print(egrad2(f2)(x,xi))
print("")
print("Function f3")
print(egrad2(f3)(x,xi))
print("")


```

which results in,

```python
Function f
[[  1.   0.   0.]
 [  1.   2.   3.]
 [  1.   4.  12.]
 [  1.   6.  27.]
 [  1.   8.  48.]
 [  1.  10.  75.]
 [  1.  12. 108.]
 [  1.  14. 147.]
 [  1.  16. 192.]
 [  1.  18. 243.]]

Function f2
[   0.    9.   48.  141.  312.  585.  984. 1533. 2256. 3177.]

Function f3
[[   0.    0.    0.    0.    0.    0.    0.    0.    0.    0.]
 [   9.    9.    9.    9.    9.    9.    9.    9.    9.    9.]
 [  48.   48.   48.   48.   48.   48.   48.   48.   48.   48.]
 [ 141.  141.  141.  141.  141.  141.  141.  141.  141.  141.]
 [ 312.  312.  312.  312.  312.  312.  312.  312.  312.  312.]
 [ 585.  585.  585.  585.  585.  585.  585.  585.  585.  585.]
 [ 984.  984.  984.  984.  984.  984.  984.  984.  984.  984.]
 [1533. 1533. 1533. 1533. 1533. 1533. 1533. 1533. 1533. 1533.]
 [2256. 2256. 2256. 2256. 2256. 2256. 2256. 2256. 2256. 2256.]
 [3177. 3177. 3177. 3177. 3177. 3177. 3177. 3177. 3177. 3177.]]
```

The shape of the first two outputs matches my expectation. However, the shape of the third output does not. I'd expect the shape of the third output to be a shape (10,) array like,

```python
[   0.    9.   48.  141.  312.  585.  984. 1533. 2256. 3177.]
```

In other words, I'd expect the outputs of function 2 and function 3 to be identical. Have I done something incorrectly when creating the egrad functions, or are the shapes of the outputs to be expected?

Thank you

**Comment by mattjj** [comment]
Always great to hear from Autograd users!

By `egrad` you mean [`elementwise_grad`](https://github.com/HIPS/autograd/blob/c6f630a5ec18bd30f1485bc0dbbccb8664c77510/autograd/differential_operators.py#L32), right? If so, you can implement it just like how Autograd does, just applying a `vjp` to an all-ones vector:

```python
#Import JAX
from jax.config import config
config.update('jax_enable_x64', True)

import jax
import jax.numpy as np

# Create simple data
x = np.arange(10,dtype=np.float64)
xi = np.ones(3)

A = np.ones((10,3))

# Create three functions, increasing in complexity
f = lambda x: np.array([x,x**2,x**3]).T
f2 = lambda x,xi: x*np.dot(f(x),xi)
f3 = lambda x,xi: x*np.dot(f(x),xi)+np.dot(A,xi)

def egrad(g):
  def wrapped(x, *rest):
    y, g_vjp = jax.vjp(lambda x: g(x, *rest), x)
    x_bar, = g_vjp(np.ones_like(y))
    return x_bar
  return wrapped

# Print results
print("Function f")
print(egrad(f)(x))
print("")
print("Function f2")
print(egrad(f2)(x,xi))
print("")
print("Function f3")
print(egrad(f3)(x,xi))
print("")
```

(This version assumes there's a single array output, but that's easy to generalize.)

That'll act just like Autograd's `elementwise_grad`, but since these functions don't act elementwise, this wasn't the original intended usage of `elementwise_grad`! That is, we're just applying a VJP to an all-ones vector, giving us the sum of the rows of the Jacobian; when the Jacobian is diagonal then that gives the diagonal elements, but it won't give the diagonal elements of the Jacobian for general functions.

(We can also use `jax.vmap` together with `jax.grad` or `jax.jvp` to efficiently compute the diagonal Jacobian elements of functions with diagonal Jacobians, but that's a different technique than `elementwise_grad`.)

WDYT?

**Comment by leakec** [comment]
This is perfect, and exactly what I was looking for; thank you! 

I know that this was not the original intention of elementwise_grad, but it worked perfectly for what I needed. I loved autograd, and am really excited about JAX as well. A big thanks to everyone who has made it possible :)

This completely solves my issue, so I will close it.

**Comment by mattjj** [comment]
I think the autodiff cookbook may be a pretty canonical place (though I'm interested if folks have other ideas). That seems like a good place for "recipes".

IIUC we could say:
1. if your function is already vectorized/mapped, use vjp-with-ones
2. if your function is not yet vectorized/mapped, use vmap-of-grad

Does that sound right?
